universal-application-tool-0.0.1/test/controllers/applicant/ApplicantProgramsControllerTest.java

```diff
@@ -0,0 +1,68 @@
+package controllers.applicant;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static play.mvc.Http.Status.OK;
+import static play.test.Helpers.contentAsString;
+
+import models.Program;
+import org.junit.Before;
+import org.junit.Test;
+import play.mvc.Result;
+import repository.WithPostgresContainer;
+
+public class ApplicantProgramsControllerTest extends WithPostgresContainer {
+
+  private ApplicantProgramsController controller;
+
+  @Before
+  public void setupController() {
+    controller = instanceOf(ApplicantProgramsController.class);
+  }
+
+  @Test
+  public void index_withNoPrograms_returnsEmptyResult() {
+    Result result = controller.index(1L).toCompletableFuture().join();
+
+    assertThat(result.status()).isEqualTo(OK);
+    assertThat(result.contentType()).hasValue("text/html");
+    assertThat(result.charset()).hasValue("utf-8");
+    assertThat(contentAsString(result)).contains("Programs");
+  }
+
+  @Test
+  public void index_withPrograms_returnsAllPrograms() {
+    insertProgram("one");
+    insertProgram("two");
+
+    Result result = controller.index(1L).toCompletableFuture().join();
+
+    assertThat(result.status()).isEqualTo(OK);
+    assertThat(contentAsString(result)).contains("one");
+    assertThat(contentAsString(result)).contains("two");
+  }
+
+  @Test
+  public void index_withProgram_includesApplyButtonWithRedirect() {
+    Program program = insertProgram("program");
+
+    Result result = controller.index(1L).toCompletableFuture().join();
+
+    assertThat(result.status()).isEqualTo(OK);
+    assertThat(contentAsString(result))
+        .contains(routes.ApplicantProgramsController.edit(1L, program.id).url());
+  }
+
+  @Test
+  public void edit_returnsTemporaryString() {
+    Result result = controller.edit(123L, 456L).toCompletableFuture().join();
+
+    assertThat(result.status()).isEqualTo(OK);
+    assertThat(contentAsString(result)).isEqualTo("Applicant 123 chose program 456");
+  }
+
+  private static Program insertProgram(String name) {
```

Comment:
suggestion (non-blocking): Would it make sense to share this helper method with the [same one in AdminProgramsController](https://github.com/seattle-uat/universal-application-tool/blob/main/universal-application-tool-0.0.1/test/controllers/admin/AdminProgramControllerTest.java#L148)?

I'm not sure if it's stylistically desirable in Java, but in TypeScript, we'll often have separate test utility files that have helpers like this. This method is pretty simple, but it might make sense to break it out in a future PR, as our tests add more common database interactions.

Reply:
That's a really good idea, we can have test utilities. I filed https://github.com/seattle-uat/universal-application-tool/issues/234